src/openms/source/QC/MzCalibration.cpp

```diff
@@ -0,0 +1,193 @@
+#include <OpenMS/MATH/MISC/MathFunctions.h>
+
+using namespace std;
+ 
+namespace OpenMS
+{
+  MzCalibration::MzCalibration() : mz_raw_{}, mz_ref_{}, no_mzml_(false), name_("MzCalibration")
+  {}
+  // find original m/z Value, set meta value "mz_raw" and set meta value "mz_ref"
+  void MzCalibration::compute(FeatureMap& features, const MSExperiment& exp)
+  {
+    if (features.empty())
+    {
+      LOG_WARN << "The FeatureMap is empty.\n";
+    }
+    if (exp.empty())
+    {
+      no_mzml_ = true;
+      LOG_WARN << "Metric MzCalibration received an empty mzml file. Only reporting uncalibrated mz error.\n";
+    }
+    else
+    {
+      no_mzml_ = false;
+      // check for Calibration
+      auto is_not_elem = [](const boost::shared_ptr<const OpenMS::DataProcessing> &dp)
+      {
+        return (dp->getProcessingActions().count(DataProcessing::CALIBRATION) == 0);
+      };
+      if (all_of(exp[0].getDataProcessing().begin(), exp[0].getDataProcessing().end(), is_not_elem))
+      {
+        no_mzml_ = true;
+        LOG_WARN << "Metric MzCalibration received an mzml file which did not undergo InternalCalibration. Only reporting uncalibrated mz error.\n";
+      }
+    }
+
+    // set meta values for the first hit of all PeptideIdentifications of all features
+    for (Feature& feature : features)
+    {
+      if (feature.getPeptideIdentifications().empty())
+      {
+        LOG_WARN << "A Feature is empty.\n";
+        continue;
+      }
+
+      for (PeptideIdentification& peptide_ID : feature.getPeptideIdentifications())
+      {
+        if (!peptide_ID.hasRT())
+        {
+          LOG_WARN << "A PeptideIdentification has no retention time value.\n";
+          continue;
+        }
+
+        if (peptide_ID.hasRT())
+        {
+          if (peptide_ID.getHits().empty())
+          {
+            LOG_WARN << "A PeptideHit is empty.\n";
+            continue;
+          }
+          mz_ref_ = peptide_ID.getHits()[0].getSequence().getMonoWeight(OpenMS::Residue::Full, peptide_ID.getHits()[0].getCharge())
+                    / peptide_ID.getHits()[0].getCharge();
+          if (no_mzml_)
+          {
+            peptide_ID.getHits()[0].setMetaValue("uncalibrated_mz_error_ppm", Math::getPPM(peptide_ID.getMZ(), mz_ref_));
+          }
+          else
+          {
+            mz_raw_ = getMZraw_(peptide_ID.getRT(), exp);
+            peptide_ID.getHits()[0].setMetaValue("mz_raw", mz_raw_);
+            peptide_ID.getHits()[0].setMetaValue("mz_ref", mz_ref_);
+            peptide_ID.getHits()[0].setMetaValue("uncalibrated_mz_error_ppm", Math::getPPM(mz_raw_, mz_ref_));
+            peptide_ID.getHits()[0].setMetaValue("calibrated_mz_error_ppm", Math::getPPM(peptide_ID.getMZ(), mz_ref_));
+          }
+        }
+      }
+    }
+    // set meta values for the first hit of all unasssigned PeptideIdentifications
+    for (PeptideIdentification& unassigned_ID : features.getUnassignedPeptideIdentifications())
```

Comment:
this looks like duplicated code. Can you move that into a separate function which accepts a PI